pmount-hal.1
.TH pmount-hal 1 "November 15, 2004" "Martin Pitt"

.SH NAME
pmount-hal \- HAL-aware wrapper around pmount

.SH SYNOPSIS

.B pmount-hal
.I udi
[
.I pmount options
]

.SH DESCRIPTION

.B pmount
("policy mount") is a wrapper around the standard mount program which
permits normal users to mount removable devices without a matching /etc/fstab
entry.

.B pmount-hal 
extends pmount by making it work together with hal (Hardware Abstration Layer).
.B pmount-hal
will ask hal about values for certain mount flags (like
.B noatime\fR and \fBasync\fR) and the prefered mount point name and
pass it to
.B pmount\fR, thus respecting the configured device storage policy.

The first parameter (\fIudi\fR) must be the HAL UDI of the device to
be mounted. All further parameters are passed to pmount unchanged.

.SH SEE ALSO

.BR pmount (1),
.BR hald (8),

.SH AUTHOR
.B pmount 
is developed by Martin Pitt <martin.pitt@canonical.com>.


CHANGES
0.9.2
-----
- Switched version control to bazaar-ng (http://www.bazaar-ng.org);
  publicly accessible branch is at
  http://people.ubuntu.com/~pitti/bzr/pmount/
- pmount-hal: ignore invalid file systems, autodetect in this case
- Mark JFS and iso9660 as iocharset-capable
- pmount-hal: read volume.policy.mount_option.umask from hal.
- Fix "make dist" to work properly with the bazaar-ng metadata

0.9.1
-----
- Makefile: fix "make dist" tarball generation
- pmount: use async by default to not destroy flash memory

0.9
---
- pmount-hal: ported to new hal 0.5 and dbus 0.32 API
- integrate cryptsetup (with LUKS extension) if device is encrypted
  with dmcrypt and LUKS metadata
- pmount: mount vfat with shortname=winnt option
- pmount-hal: use volume.policy.mount_filesystem instead of
  volume.fstype
- always build the POT file, delete it on "make clean"
- Makefile: added "dist" target

0.8
---
- unbreak locking, this created ".created_by_pmount" stamps in the
  locking directories
- accept pumount by label name
- accept pmount by mount path (only in /media)
- added --umask option
- renamed template.pot to pmount.pot to match translation domain
- support a whitelist in /etc/pmount.allow

0.7.2
-----
- updated translations from Rosetta: ca, es_ES
- added new translations from Rosetta: cs, es, fr, hr, it_IT, nb

0.7.1
-----
- pmount-hal: consider a directory empty if it only contains a
  .created-by-pmount file; this makes it work with unmounting using
  umount instead of pumount

0.7
---
- support --exec/-e option
- rewrote pmount-hal in C, now it runs 10 times faster

0.6
---
- pmount-hal: use hal-supplied file system type (if present) to speed up
  mounting and to avoid bogus error messages in kernel log
- mount vfat with 'quiet' option
- support pmount/pumount'ing mount points
- added Catalan translation, thanks to Eduard Duran
- added Spanish translation, thanks to Gustavo de Lama and Jorge
  González González

0.5.1
-----
- pmount: do not allow empty media labels; this allowed to mount a
  device directly under /media if /media was empty

0.5
---
- use stamp files to mark directories created by pmount and only
  remove a mount point directory if it contains a stamp file
- use the given device name as default mount point name instead of the
  symlink-resolved one
- i18n support
- added German translation

0.4.4
-----
- pmount-hal: resolve symlinks before checking mount point directories
- use umask "077" instead of "007" for VFAT and NTFS volumes; this allows e. g.
  to use ssh keys on USB sticks
- if a mount with given iocharset fails, try again without; there were reports
  of failed vfat mounts when giving iocharset

0.4.3
-----
- pmount-hal: when checking for already existing mount points, also
  accept existing, but empty directories (if nothing is mounted at
  them)

0.4.2
-----
- pumount: allow to unmount any device if called as root

0.4.1
-----
- consistently use the realpath() of devices for all operations but
  transparent mount fallback

0.4
---
- pmount: added --noatime option
- Makefile: support PREFIX
- Makefile: install manapges, too
- Makefile: add uninstall target
- added pmount-hal (incl. manpage)

0.3
---
- pmount: added option -a/--async for asynchronous mounting
- pmount: added option -t for selecting a file system
- added NTFS support
- pmount: added option -c/--charset for selecting the iocharset;
  default to 'utf8' when pmount is executed in an UTF-8 locale,
  otherwise default to mount default
- pmount: do not fail if a device is handled in /etc/fstab, but a
  label is supplied; merely print a warning

0.2.2
-----
- mount CD-ROM file systems without umask parameter (this does not work)
- added and documented dynamic debug option -d, use dynamic debug output
  everywhere
- pumount: now use getopt_long() for option parsing, added -d and -h

0.2.1
-----
- version 0.2 was not published, had a major policy hole: in some cases it
  allowed to mount partitions on fixed disks
- policy.c: rewrote find_sysfs_device() to check minor device number, too

0.2 [DO NOT USE]
----------------
- major refactorization to clean up the code
- consolidated error return codes
- pmount: added --lock/--unlock commands
- device is regarded as removable also if /sys/block/drive/removable is 1
- mount devices with umask=007

0.1
---
- fixed /dev/ path checking: now use the real path of the device node, not the
  path supplied on the command line
- pmount: don't try to mount all file systems with the uid and gid options,
  this does not work on ext3, ext2, reiserfs, xfs, and jfs; try two separate
  mount commands now
- pumount: added option -l to support lazy unmounting
- pumount: changed policy: 
  * device must be mounted by the same uid that attempts to pumount
  * device node does not need to exist and be removeable any more (because the
    device node cannot be checked any more if device was already ripped out)

0.0.8
-----
- use root privileges also for checking that the existing mount point is empty,
  since normal user might not be allowed to read the directory

0.0.7
-----
- replaced all str[n]cat() and str[n]cpy() with snprintf() calls; thanks to
  Matt Zimmerman for the patch

0.0.6
-----
- added support for CD-ROM file systems (udf and iso9660)
- require mount points to be empty before mount

0.0.5
-----
- now also handle devices in /etc/fstab, transparently call mount/umount in
  this case

0.0.4
-----
- drop effective user id to normal user until root is really needed
- added proper online help
- added manpages

0.0.3
-----
- pumount: added similar suid check 
- pumount: delete mount point after umounting
- determine real path of devices before checking them
- pmount: mount with proper uid and gid options

0.0.2
-----
- pmount: added check if really suid, print an error if not
- pmount: set the real user id to root before executing mount (mount
  refuses to mount a device if real uid != root)

0.0.1
-----
- initial release


pmount-hal.c
/**
 * pmount-hal.c - pmount wrapper that uses HAL to get additional information
 *
 * Author: Martin Pitt <martin.pitt@canonical.com>
 * (c) 2005 Canonical Ltd.
 * 
 * This software is distributed under the terms and conditions of the 
 * GNU General Public License. See file GPL for the full text of the license.
 */

#include <stdio.h>
#include <stdlib.h> /* calloc */
#include <limits.h> /* PATH_MAX */
#include <unistd.h> /* execvp */
#include <string.h>
#include <dirent.h>
#include <sys/stat.h>
#include <locale.h>
#include <libintl.h>
#include <libhal.h>

#include "policy.h"
#include "fs.h"

/* gettext abbreviation */
#define _(String) gettext(String)

void help() {
    puts( _(
"pmount-hal - execute pmount with additional information from hal\n\n"
"Usage: pmount-hal <hal UDI> [pmount options]\n\n"
"This command mounts the device described by the given UDI using pmount. The\n"
"file system type, the volume storage policy and the desired label will be\n"
"read out from hal and passed to pmount."));
}

int empty_dir( const char* dirname )
{
    DIR* dir;
    struct dirent* dirent;

    if( !( dir = opendir( dirname ) ) )
        return 1;

    while ( ( dirent = readdir( dir ) ) ) {
        if( strcmp( dirent->d_name, "." ) && 
            strcmp( dirent->d_name, ".created_by_pmount" ) && 
            strcmp( dirent->d_name, ".." ) ) {
            closedir( dir );
            return 0;
        }
    }

    closedir( dir );
    return 1;
}

/* Return 0 if mntpt is nonempty and/or already mounted, 1 otherwise. */
int valid_mntpt( const char* mntpt )
{
    struct stat st;
    char realp[PATH_MAX];

    if( stat( mntpt, &st ) )
        return 1; /* does not exist */

    /* resolve symlinks */
    if( S_ISLNK( st.st_mode ) ) {
        if( !realpath( mntpt, realp ) )
            return 0;
        mntpt = realp;
        if( stat( mntpt, &st ) )
            return 0; /* invalid link */
    }
    
    if( !S_ISDIR( st.st_mode ) )
        return 0;

    if( !empty_dir( mntpt ) )
        return 0;

    if( fstab_has_mntpt( "/proc/mounts", mntpt, NULL, 0 ) ||
        fstab_has_mntpt( "/etc/mtab", mntpt, NULL, 0 ) )
        return 0;

    return 1;
}

void get_free_label( const char* label, char* result, size_t result_size )
{
    char mntpt[PATH_MAX];
    snprintf( mntpt, sizeof(mntpt), "/media/%s", label );
    int n;

    if( valid_mntpt( mntpt ) ) {
        strncpy( result, label, result_size-1 );
        result[result_size-1] = 0;
        return;
    }

    for( n = 1; n > 0; ++n ) {
        snprintf( mntpt, sizeof(mntpt), "/media/%s-%i", label, n );
        if( valid_mntpt( mntpt ) ) {
            snprintf( result, result_size, "%s-%i", label, n );
            return;
        }
    }
}

void exec_pmount( const char* device, const char* fstype, const char* label,
        dbus_bool_t sync, dbus_bool_t noatime, const char* umask, int addargc,
        const char* const* addargv ) 
{
    const char** argv = (const char**) calloc( sizeof( const char* ), addargc+12 );
    int argc = 0;
    char freelabel[PATH_MAX];
    int i;

    argv[argc++] = "pmount";

    if( fstype ) {
        argv[argc++] = "-t";
        argv[argc++] = fstype;
    }

    if( sync )
        argv[argc++] = "--sync";
    if( noatime )
        argv[argc++] = "--noatime";

    if( umask ) {
        argv[argc++] = "--umask";
        argv[argc++] = umask;
    }

    for( i = 0; i < addargc; ++i )
        argv[argc++] = addargv[i];

    argv[argc++] = device;
    if( label ) {
        get_free_label( label, freelabel, sizeof( freelabel ) );
        argv[argc++] = freelabel;
    }
    argv[argc] = NULL;

    execvp( argv[0], (char *const *) argv );
    fputs( _("Error: could not execute pmount\n"), stderr );
    exit( 1 );
}

int
main( int argc, const char** argv ) 
{
    LibHalContext *hal_ctx;
    DBusError error;
    DBusConnection *dbus_conn;
    dbus_bool_t sync = FALSE, noatime = FALSE;
    const char* udi;
    char* device = NULL;
    char* label = NULL;
    char* fstype = NULL;
    char *umask = NULL;

    /* initialize locale */
    setlocale( LC_ALL, "" );
    bindtextdomain( "pmount", NULL );
    textdomain( "pmount" );

    if( argc < 2 ) {
        help();
        return 0;
    }

    udi = argv[1];

    /* initialize hal connection */
    dbus_error_init( &error );
    dbus_conn = dbus_bus_get( DBUS_BUS_SYSTEM, &error );
    if( dbus_conn == NULL ) {
        fprintf( stderr, _("Error: could not connect to dbus: %s: %s\n"), error.name, error.message );
        return 1;
    }

    hal_ctx = libhal_ctx_new();
    if( !hal_ctx ) {
        fprintf( stderr, "Error: libhal_ctx_new\n" );
        return 1;
    }

    if( !libhal_ctx_set_dbus_connection( hal_ctx, dbus_conn ) ) {
        fprintf( stderr, "Error: libhal_ctx_set_dbus_connection: %s: %s\n", error.name, error.message );
        return 1;
    }
    if( !libhal_ctx_init( hal_ctx, &error ) ) {
        fprintf(  stderr, "Error: libhal_ctx_init: %s: %s\n", error.name, error.message );
        return 1;
    }

    /* get all interesting properties */
    if( !libhal_device_exists( hal_ctx, udi, &error ) ) {
        fprintf( stderr, _("Error: UDI '%s' does not exist\n"), udi );
        return 1;
    }

    if( libhal_device_property_exists( hal_ctx, udi, "block.device", &error ) )
        device = libhal_device_get_property_string( hal_ctx, udi, "block.device", &error );
    if( !device ) {
        fprintf( stderr, _("Error: given UDI is not a mountable volume\n") );
        return 1;
    }

    if( libhal_device_property_exists( hal_ctx, udi, "volume.policy.desired_mount_point", &error ) )
        label = libhal_device_get_property_string( hal_ctx, udi, "volume.policy.desired_mount_point", &error );
    if( libhal_device_property_exists( hal_ctx, udi, "volume.policy.mount_filesystem", &error ) ) {
        fstype = libhal_device_get_property_string( hal_ctx, udi, "volume.policy.mount_filesystem", &error );
        /* ignore invalid file systems */
        if (fstype && !get_fs_info(fstype)) {
            free (fstype);
            fstype = NULL;
        }
    }
    if( libhal_device_property_exists( hal_ctx, udi, "volume.policy.mount_option.sync", &error ) )
        sync = libhal_device_get_property_bool( hal_ctx, udi, "volume.policy.mount_option.sync", &error );
    if( libhal_device_property_exists( hal_ctx, udi, "volume.policy.mount_option.noatime", &error ) )
        noatime = libhal_device_get_property_bool( hal_ctx, udi, "volume.policy.mount_option.noatime", &error );
    if( libhal_device_property_exists( hal_ctx, udi, "volume.policy.mount_option.umask", &error ) )
        umask = libhal_device_get_property_string( hal_ctx, udi, "volume.policy.mount_option.umask", &error );

    /* shut down hal connection */
    libhal_ctx_shutdown( hal_ctx, &error );
    libhal_ctx_free( hal_ctx );
    dbus_connection_disconnect( dbus_conn );
    dbus_connection_unref( dbus_conn );

    /* go */
    exec_pmount( device, fstype, label, sync, noatime, umask, argc-2, argv+2 );

    return 0;
}

